var/spack/repos/builtin/packages/cmake/package.py

```diff
@@ -141,6 +146,25 @@ class Cmake(Package):
 
     phases = ['bootstrap', 'build', 'install']
 
+    @staticmethod
+    def determine_spec_details(prefix, exes_in_prefix):
+        exe_to_path = dict(
+            (os.path.basename(p), p) for p in exes_in_prefix
+        )
+        if 'cmake' not in exe_to_path:
+            return None
+
+        spec = spack.spec.Spec('cmake')
+
+        cmake = spack.util.executable.Executable(exe_to_path['cmake'])
+        output = cmake('--version', output=str)
+        if output:
+            match = re.search(r'version\s+(\S+)', output)
+            if match:
+                version_str = match.group(1)
+            spec.versions = [Version(version_str)]
```

Comment:
I think this should read:
```suggestion
            spec.versions = Version(version_str)
```
for the version to be seen as concrete.

Reply:
This is strange: I generally like to build up a Spec object and then stringify it. When I did as you suggest using a nonstandard version:

`spec.versions = Version('1.foo')`

I got a strange error:

```
>       assert single_entry.spec == Spec('cmake@1.foo')

lib/spack/spack/test/cmd/external.py:47: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
lib/spack/llnl/util/lang.py:228: in <lambda>
    (s is o) or (o is not None and s._cmp_key() == o._cmp_key()))
lib/spack/spack/version.py:91: in coercing_method
    ca, cb = coerce_versions(a, b)
lib/spack/spack/version.py:68: in coerce_versions
    check_type(tb)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

t = <type 'int'>

    def check_type(t):
        if t not in order:
>           raise TypeError("coerce_versions cannot be called on %s" % t)
E           TypeError: coerce_versions cannot be called on <type 'int'>
```

So I think there might be an issue for how `Spec.__str__` works. IMO either mine or your expression should result in a valid string. I don't think we technically have an API for constructing Specs manually, so it might be better to suggest that users construct strings that represent Specs instead of the Specs themselves.

Reply:
> I don't think we technically have an API for constructing Specs manually, so it might be better to suggest that users construct strings that represent Specs instead of the Specs themselves.

:100: on that. I think packagers would be in general more accustomed to work with spec strings that to construct a Spec object using the internal API.